### PR TITLE
go/signedexchange: Refactor request method handling

### PR DESCRIPTION
**Pull request by irori**
- Add `RequestMethod` field to the `Exchange` struct
- `NewExchange()` takes `method` as a parameter
- Fix a bug that request headers were not encoded
- Function Renamings
  - {encode,decode}Request -> {encode,decode}RequestMap
  - {encode,decode}ResponseHeaders -> {encode,decode}ResponseMap

### CODE REVIEW

File: go/signedexchange/cmd/gen-signedexchange/main.go [hunk 1]
```diff
@@ -156,7 +156,7 @@ func run() error {
 	if resHeader.Get("content-type") == "" {
 		resHeader.Add("content-type", "text/html; charset=utf-8")
 	}
-	e, err := signedexchange.NewExchange(ver, parsedUrl, reqHeader, *flagResponseStatus, resHeader, payload)
+	e, err := signedexchange.NewExchange(ver, parsedUrl, "GET", reqHeader, *flagResponseStatus, resHeader, payload)
```
Comment (hajimehoshi) [inline go/signedexchange/cmd/gen-signedexchange/main.go]: `http.MethodGet` sounds better

File: go/signedexchange/signedexchange_test.go [hunk 2]
```diff
@@ -65,15 +65,17 @@ func mustReadFile(path string) []byte {
...
 
 	// Multiple values for the same header
-	header.Add("Foo", "Bar")
-	header.Add("Foo", "Baz")
+	respHeader.Add("Foo", "Bar")
+	respHeader.Add("Foo", "Baz")
 
 	const ver = version.Version1b1
-	e, err := NewExchange(ver, u, nil, 200, header, []byte(payload))
+	e, err := NewExchange(ver, u, "GET", reqHeader, 200, respHeader, []byte(payload))
```
Comment (hajimehoshi) [inline go/signedexchange/signedexchange_test.go]: ditto

File: go/signedexchange/signedexchange.go [hunk 3]
```diff
@@ -50,14 +51,15 @@ var (
...
-
-	valueGet = []byte("GET")
 )
 
-func NewExchange(ver version.Version, uri *url.URL, requestHeaders http.Header, status int, responseHeaders http.Header, payload []byte) (*Exchange, error) {
+func NewExchange(ver version.Version, uri *url.URL, method string, requestHeaders http.Header, status int, responseHeaders http.Header, payload []byte) (*Exchange, error) {
 	if uri.Scheme != "https" {
 		return nil, fmt.Errorf("signedexchange: The request with non-https scheme %q URI can't be captured inside signed exchange.", uri.Scheme)
 	}
+	if method != "GET" && method != "HEAD" {
```
Comment (hajimehoshi) [inline go/signedexchange/signedexchange.go]: `http.MethodGet` and `http.MethodHead`

File: go/signedexchange/cmd/gen-signedexchange/main.go (hunk 1 above)
Comment (irori) [reply go/signedexchange/cmd/gen-signedexchange/main.go]: Done

File: go/signedexchange/signedexchange_test.go (hunk 2 above)
Comment (irori) [reply go/signedexchange/signedexchange_test.go]: Done

File: go/signedexchange/signedexchange.go (hunk 3 above)
Comment (irori) [reply go/signedexchange/signedexchange.go]: Done

File: go/signedexchange/signedexchange.go [hunk 4]
```diff
@@ -187,24 +186,29 @@ func (e *Exchange) decodeRequest(dec *cbor.Decoder) error {
...
 	}
-	for name, value := range e.ResponseHeaders {
-		mes = append(mes,
+	mes = encodeHeaders(mes, e.ResponseHeaders)
+	return enc.EncodeMap(mes)
+}
+
+func encodeHeaders(encoder []*cbor.MapEntryEncoder, headers http.Header) []*cbor.MapEntryEncoder {
+	for name, value := range headers {
+		encoder = append(encoder,
```
Comment (nyaxt) [inline go/signedexchange/signedexchange.go]: Make plural? `encoders` or `encs` or `es`?

Comment (irori) [reply go/signedexchange/signedexchange.go]: done

**Comment by irori** [comment]
Pushed one more commit. PTAL

- gen-signedexchange: Add `-method` option
- dump-signedexchange: Print exchange's request method

**Pull request merged**